lib/IRGen/IRGenSIL.cpp

```diff
@@ -2484,6 +2530,66 @@ void IRGenSILFunction::visitFullApplySite(FullApplySite site) {
                              subMap, &witnessMetadata, llArgs);
   }
 
+  auto *silFn = site.getInitiallyReferencedFunction();
+  if (silFn && silFn->getClangDecl() && silFn->hasIndirectFormalResults() &&
+      isa<clang::FunctionTemplateDecl>(silFn->getClangDecl())) {
+    // This function template returns a template. We can only return one thing
+    // in C++.
+    const auto &fnPtr = emission.getCallee().getFunctionPointer();
+    auto *fnReturnType = fnPtr.getFunctionType()->getReturnType();
+    // There's a good chance that the return type is no longer indirect.
+    if (!fnReturnType->isVoidTy()) {
+      // If that's the case, store the result of the function in the last
+      // argument (the indirect result pointer).
+      auto *resultArg = llArgs.claimNext();
+      Explosion newArgs;
+      size_t i = 0;
+      while (llArgs.size()) {
+        auto *paramType = fnPtr.getFunctionType()->getParamType(i++);
+        auto *argAddr = llArgs.claimNext();
+        if (paramType->isPointerTy()) {
+          newArgs.add(coerceValue(argAddr, paramType, IGM.DataLayout));
+        } else {
+          argAddr = Builder.CreateBitCast(argAddr, paramType->getPointerTo());
+          newArgs.add(Builder.CreateLoad(argAddr, IGM.getPointerAlignment()));
+        }
+      }
+      auto *result = Builder.CreateCall(fnPtr, newArgs.claimAll());
+      // If the return type is a pointer, emit a copy to the result argument.
+      // Otherwise, just cast and store the return value.
+      if (result->getType()->isPointerTy()) {
+        // Get the result type and cast both the return pointer and result
+        // argument to that storage type.
+        auto resultSILTy = site.getSubstCalleeType()
+                               ->getSingleResult()
+                               .getSILStorageInterfaceType();
+        const auto &resultTypeInfo = getTypeInfo(resultSILTy);
+        auto *storageType = resultTypeInfo.getStorageType()->getPointerTo();
+        auto *resultPtr =
+            result->getType() == storageType
+                ? result
+                : coerceValue(result, storageType, IGM.DataLayout);
+        auto *resultArgPtr =
+            resultArg->getType() == storageType
+                ? resultArg
+                : coerceValue(resultArg, storageType, IGM.DataLayout);
+        Address resultAddr = resultTypeInfo.getAddressForPointer(resultPtr);
+        Address resultArgAddr =
+            resultTypeInfo.getAddressForPointer(resultArgPtr);
+        // Then copy the returned pointer into the result argument.
+        resultTypeInfo.initializeWithCopy(*this, resultArgAddr, resultAddr,
+                                          resultSILTy, false);
+      } else {
+        auto *resultArgPtr =
+            Builder.CreateBitCast(resultArg, result->getType()->getPointerTo());
+        Builder.CreateStore(result, resultArgPtr, IGM.getPointerAlignment());
+      }
+      // Now we're done.
+      emission.clear();
```

Comment:
This seems to imply that you have unclaimed parameter values along this open-coded path.